posts/napari-rois/cosmx-rois.qmd

```diff
@@ -0,0 +1,492 @@
+to select the vertices of a given polygon. Double-clicking the last vertex will
+"close" the polygon. 
+
+:::{.callout-tip}
+I tend to keep the shapes layer inactive whenever possible. This limits the number
+of accidental vertex additions to a shape. 
+:::
+
+Here's the code. Be sure to add the shortcut definitions _after_ creating the `gem` object.
+You can change the color and edge width to any valid value. Here I'll make the 
+border yellow with a width of 100.
+
+```{.python}
+
+def find_next_roi(my_list):
+  """Finds the next ROI code in a list of ROI strings.
+
+  Args:
+      my_list: A list of strings in the format "ROIxxx" where xxx are digits.
+
+  Returns:
+      The next ROI code in the sequence, or "ROI001" if the list is empty 
+      or the next sequential code if no gap is found.
+  """
+
+  if not my_list:  # Handle empty list
+    return "ROI001"
+
+  # Extract numeric parts and convert to integers
+  numbers = [int(item[3:]) for item in my_list]
+
+  # Sort the numbers in ascending order
+  numbers.sort()
+
+  # Find the first gap in the sequence
+  for i in range(len(numbers) - 1):
+    if numbers[i + 1] - numbers[i] > 1:
+      next_number = numbers[i] + 1
+      return f"ROI{next_number:03d}"
+
+  # If no gap is found, return the next sequential number
+  next_number = numbers[-1] + 1  # Get the last number and increment
+  return f"ROI{next_number:03d}"
+
+# This keybinding makes it easier to create a new shape layer for ROIs
+@gem.viewer.bind_key('n', overwrite=True)
+def new_roi_layer(viewer):
+    # Search layers for "ROI" pattern
+    layers = viewer.layers
+    names = [x.name for x in layers]
+    names_filtered = [item for item in names if "ROI" in item]
+    new_roi_name = find_next_roi(names_filtered)
+    print(f"Creating {new_roi_name} layer.")
+    viewer.add_shapes(
+        shape_type='polygon',
+        scale = [gem.mm_per_px, gem.mm_per_px],
+        name=new_roi_name,
+        face_color='transparent',
+        edge_color='yellow',
+        edge_width=100
+    )
+
+
+```
+
+In the video below, I am quickly creating a new shapes layer using the keyboard
+combination `n + p`.
+
+{{< video ./figures/creating_rois.mov >}} 
+
+
+And now that I have four ROIs, I can save them like this: 
+
+```{.python}
+N=4
+roi_list = ['ROI'+str(i).zfill(3) for i in range(1, N+1)]
+gem.save_layers(roi_list, "path/to/output/data")
+```
+
+
+## Subsetting Based on ROIs
+
+What we would like to do now is to identify which of the cells are in each of these
+polygons. There's a method available in `napari-cosmx` to do just that. 
+
+Using the `roi_list` list we created above, we can generate a Boolean column
+in the metdata for all of the ROIs. 
+
+:::{.callout-note}
+Depending on the size of the sample, the size of the ROI(s), and your hardware, this
+process can be computationally intense. 
+:::
+
+
+```{.python}
+new_metadata = gem.layers_to_metadata(roi_list)
+
+new_metadata
```

Comment:
Can you provide an example of saving these metadata? Can you also provide an example where you replace the gem metadata with these new metadata, so that you can select these ROI cells in the widget?

Reply:
done